rfcs/reuse-insert-permission-in-action.md

```diff
@@ -0,0 +1,67 @@
+### Insert permissions and actions
+
+If an action is used for data validation, it is hard to re-use constraints
+specified through insert permissions.
+
+#### Motivation
+
+Let's consider a table with the following schema:
+
+```sql
+create table message (
+  id serial primary key,
+  content text not null,
+  channel_id integer not null references channel(id)
+)
+create table channel (..)
+create table channel_members (..)
+```
+
+Let's say you want to allow a `user` role to insert into
+`message` table if the following `check` constraint holds:
+```json
+{"channel": {"members": {"user_id": "x-hasura-user-id"}}}
+```
+i.e, a user can only post a message to a channel if they are a member of that
+particular channel. Now let's say you want to do some sort of validation on the
+content of the message (maybe through some external service), this is a problem
+that would broadly fall under the 'data validation' category (data cannot be
+validated with graphql-engine's permissions or through Postgres's check
+constraints).
+
+We recommend actions in such cases, we would suggest the developer to remove
+the permission for `message` table on `user` and then do the entire validation
+in the webhook for the action. However the webhook now has to do more than
+validation of the message's content - it also has to check this constraint
+`{"channel": {"members": {"user_id": "x-hasura-user-id"}}}` which was part of
+insert permissions earlier. This is inconvenient and the reusability of
+hasura's permission system is lost.
+
+Why should the insert permission be removed? Because if the insert permission
+is defined, the `insert_message` mutation is generated which can be directly
+used by anyone with the `user` role and thus bypassing the validation of the
+`message` content.
+
+This problem is originally reported
+[on discord](https://discordapp.com/channels/407792526867693568/516590536875048990/685462912420282546).
+
+#### Proposed solution
+
+Introduce a new field in insert permission called `admin_only`, this
```

Comment:
How is the solution affected by https://github.com/hasura/graphql-engine/blob/233ab645993ec08e981ca55d4bc8414a01bbd944/rfcs/disable-query-and-subscription-root-fields.md

Does it make sense to hide fields under `mutation` also? In that case, we do not need `admin_only` in the permission but we will need a header like `x-hasura-include-hidden-fields: true` or something?